### PR TITLE
AST: avoid unnecessarily calling getStartLoc() to avoid exponential growth of the stack trace

### PR DESCRIPTION
**Pull request by nkcsgexi**
Issuing multiple getStartLoc() from sub-expression can exponentially grow the stack trace.
When the expression under analysis is complex enough, this could be a user-noticeable hang.
This patch fixes UnresolvedDotExpr::getStartLoc() by 'refactoring' the result of SubExpr->getStartLoc() to a local variable.

rdar://52982457

### CODE REVIEW

File: test/IDE/coloring_52982457.swift [hunk 1]
```diff
@@ -0,0 +1,115 @@
+// RUN: %target-swift-ide-test -syntax-coloring -source-filename %s | %FileCheck %s
```
Comment (akyrtzi) [inline test/IDE/coloring_52982457.swift]: Could you add a comment at the top of the test to make it clear this is to check that the test case doesn't take a long time to finish?

**Review by akyrtzi — APPROVED** [review]
It would be good to add a clarifying comment for the purpose of the test case, otherwise LGTM!

File: test/IDE/coloring_52982457.swift (hunk 1 above)
Comment (nkcsgexi) [reply test/IDE/coloring_52982457.swift]: Done

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Pull request merged**